lib/beautiful/gtk.lua

```diff
@@ -0,0 +1,169 @@
+  --result.font_size = font:get_size()/1024
+  result.font = font:to_string()
+
+  result = join(result, read_gtk_color_properties_from_widget(
+    window, {
+      bg_color="background-color",
+      fg_color="color",
+    }
+  ))
+  result = join(result, read_gtk_color_properties_from_widget(
+    Gtk.Entry(), {
+      base_color="background-color",
+      text_color="color",
+    }
+  ))
+  result = join(result, read_gtk_color_properties_from_widget(
+    Gtk.ToggleButton(), {
+      selected_bg_color="background-color",
+      selected_fg_color="color",
+    }
+  ))
+  local button = Gtk.Button()
+  result = join(result, read_gtk_color_properties_from_widget(
+    button, {
+      button_bg_color="background-color",
+      button_fg_color="color",
+    }
+  ))
+  local button_style_context = button:get_style_context()
+  for result_key, style_context_property in pairs({
+    border_radius="border-radius",
+    border_width="border-top-width",
+  }) do
+    local state = style_context:get_state()
+    local property = button_style_context:get_property(style_context_property, state)
+    result[result_key] = property.value
+    property:unset()
+  end
+
+  for _, color_data in ipairs({
+    {"bg_color", "theme_bg_color"},
+    {"fg_color", "theme_fg_color"},
+    {"base_color", "theme_base_color"},
+    {"text_color", "theme_text_color"},
+    {"selected_bg_color", "theme_selected_bg_color"},
+    {"selected_fg_color", "theme_selected_fg_color"},
+    --
+    {"tooltip_bg_color", "theme_tooltip_bg_color", "bg_color"},
+    {"tooltip_fg_color", "theme_tooltip_fg_color", "fg_color"},
+    {"osd_bg_color", "osd_bg", "tooltip_bg_color"},
+    {"osd_fg_color", "osd_fg", "tooltip_fg_color"},
+    {"osd_border_color", "osd_borders_color", "osd_fg_color"},
+    {"menubar_bg_color", "menubar_bg_color", "bg_color"},
+    {"menubar_fg_color", "menubar_fg_color", "fg_color"},
+    --
+    {"button_bg_color", "button_bg_color", "bg_color"},
+    {"button_fg_color", "button_fg_color", "fg_color"},
+    {"header_button_bg_color", "header_button_bg_color", "menubar_bg_color"},
+    {"header_button_fg_color", "header_button_fg_color", "menubar_fg_color"},
+    --
+    {"wm_bg_color", "wm_bg", "menubar_bg_color"},
+    {"wm_border_focused_color", "wm_border_focused", "selected_bg_color"},
+    {"wm_title_focused_color", "wm_title_focused", "menubar_bg_color"},
+    {"wm_icons_focused_color", "wm_icons_focused", "menubar_fg_color"},
+    {"wm_border_unfocused_color", "wm_border_unfocused", "wm_bg_color"},
+    {"wm_title_unfocused_color", "wm_title_unfocused", "menubar_bg_color"},
+    {"wm_icons_unfocused_color", "wm_icons_unfocused", "menubar_fg_color"},
+  }) do
+    local result_key, style_context_key, fallback_key = unpack(color_data)
+    result[result_key] = lookup_gtk_color_to_hex(style_context, style_context_key) or
+      (result[result_key] ~= "#000000" and result[result_key] or result[fallback_key] or result[result_key])
```

Comment:
Uhm, replacing the actual values in the expression above with `a`, `b` etc, the above does
```
a = b or (a ~= "#000000" and a or c or a)
```
I don't know what is so special about black, so perhaps a comment should be added to explain this (*). Anyway, I want to try to simplify the expression above. Is the following easier to read? Is it equivalent to the original expression? (I'm not quite sure, which is part of why I think it should be simplified):
```
a = b or (a == "#000000" and c) or a
```
Well, no... the above is different when `result[result_key]` (aka `a`) is `nil` to begin with...
How about
```lua
local value
value = lookup_gtk_color_to_hex(style_context, style_context_key)
value = value or (result[result_key] ~= "#000000" and result[result_key])
value = value or result[fallback_key]
if value then
    result[result_key] = value
end
```
Again, I am again not sure if the above is equivalent or better... Let's wait for what @Elv13 and @blueyed say.

(*): If this black happens to be the result of `read_gtk_color_properties_from_widget` when the color property is not found: Why is that code so careful in checking for `nil` all the time?

Reply:
null color property returns #000000, null stylecontext returns null

i can rewrite it in multiline way if it's easier to read

Reply:
> null color property returns #000000, null stylecontext returns null

What exactly do you mean here? How/where would I get a null color property or a null stylecontext? I tried the following and neither gave me `#000000`, but instead `nil`:
```
lua -e 'gtk = require("lgi").Gtk sc = gtk.Window{}:get_style_context() print(sc:get_property("foo", sc:get_state()).value)
lua -e 'gtk = require("lgi").Gtk sc = gtk.Window{}:get_style_context() print(sc:lookup_color("foo"))'
```

Reply:
in your first example you're trying to retrieve `non-existing` property, not proverty with `null` value

i was able to reproduce it with `button`'s `background-color` in the themes where no background-color is set but `background: <gradient>` or just a pixmap

Reply:
Ah, okay. I haven't found were this is set in the code, but I found it in the docs: https://developer.gnome.org/gtk3/stable/chap-css-properties.html#id-1.5.3.3.34
The initial value for `background-color` is `rgba(0, 0, 0, 0)`. And apparently nothing else has this default initial value. This means two things:
- This default value should only be ignored for `background-color`, not for all fields.
- This default value likely should not be ignored at all...? Since apparently that is indeed the color that will be used by GTK here, no?

Oh, okay, I already found it in the code: https://sources.debian.org/src/gtk+3.0/3.22.26-2/gtk/gtkcssstylepropertyimpl.c/#L1114

Oh and (sorry, unrelated): There seems to be a `font-size` property: https://sources.debian.org/src/gtk+3.0/3.22.26-2/gtk/gtkcssstylepropertyimpl.c/#L1075

This all still "feels" weird....

Reply:
actually after uncommenting `.alpha` part of code it should be fine -- `#00000000` is fully transparent color

Reply:
also thanks, good idea re: font-size/font-family